### PR TITLE
Feature global state

### PR DESCRIPTION
**Pull request by RedSkotina**
Naive realisation of global state feature with backtracking support.
Idea taken from https://github.com/PuerkitoBio/pigeon/issues/7 
Implementation:
Struct current contain member `state` with type `statedict` where `type statedict map[string]interface{}`. This member can directly referenced in action or predicate codeblocks `{ c.state["some"] = 1 }` . 

Added testcase  "test/state" for test of state backtracking.
Added example "example/indenting" with implementation of easy whitespace significant language.

Action codeblocks using access to `state` are "impure" elements and cant be memoized !?(need tests)

### CODE REVIEW

**Comment by breml** [comment]
Hi @RedSkotina 
First of all, thanks for your PR and sorry for the long delay until this first reaction. As you might have noticed, I took over the maintenance of pigeon some weeks ago. Until now, I was busy with some other changes, but now I would like to seriously evaluate your PR.
Its now more than 9 months since you provided this PR and therefore I first want to ask you, if you are still interested in this PR and if you are willing to work on this with me to get it finally merged into pigeon.
Please drop me a short note. Thanks.

**Comment by RedSkotina** [comment]
I ready work on PR  , but it takes considerable time switching back into this codebase.

**Comment by breml** [comment]
@RedSkotina That is good news.

First of all, I would like to understand your reasoning behind the decision to add the new language element state codeblock. Why is it needed? Could the access to the global state also be solved in a different way (e.g. I implemented the global store feature by adding the global store map to `c *current`, see https://github.com/mna/pigeon/pull/23)?

CC: @jlozano

**Comment by RedSkotina** [comment]
@breml As far as I can remember and see, your global state implementation does not allow you to roll back the state. I it important for easy and correct Significant Whitespace Parsing.
I was based in my work on a syntax similar to https://github.com/otac0n/Pegasus/wiki/Significant-Whitespace-Parsing .

**Comment by breml** [comment]
@RedSkotina Thanks for the link, I will look into this. Do you found other PEG implementations, which have this global state feature as well?

**Comment by RedSkotina** [comment]
@breml  Maybe peg.js. I see proposals but dont understand current proposal state of "global state"
https://github.com/pegjs/pegjs/issues/45
https://github.com/pegjs/pegjs/issues/217

**Comment by breml** [comment]
@RedSkotina I looked at Pegasus as well as at the two PR for peg.js and I did some googling my self. So far I could not find an other PEG implementation (beside of Pegasus), which does provide this state language element. There are PR/suggestions for peg.js, but non of them have been merged. Also I could not find any paper, which suggests to add such a state element to PEG. This makes me cautious, because adding a new language element is not something I want to take lightly.

This being said, the possibility to have a state in the parser, on which a rollback is performed, in case of a back tracking, is something, which I think is valuable. So my question is, if it would also work to provide such a state (with the automatic rollback), which is accessed in action, and code and not code expressions the same way as the global store is accessed. Do you see any problems with such an approach?

Additionally to the new state element, there are two things int the current PR, which I do not like:

* state is a package global variable. With this it is not possible to have multiple instances of a generated parser to run concurrent.
* The state is kind of stored multiple times. This is also the reason, why upon restore, the state has to be copied twice.

In the final solution, those two issues have to be resolved.

**Comment by RedSkotina** [comment]
@breml i found this
https://pdos.csail.mit.edu/papers/packrat-parsing:ford-ms.pdf
page 53 ,  3.2.5 Packrat Parsing with State
Here described pros and cons of parsing with state.
I not understand your question. Can you rephrase?
About package global variable. I do it only for copy syntaxis of pegasus. I did not want to enter new parameters in the action function to avoid complicating the parser. So it was just a POC.
Double copying is done because I wanted to have a read only copy of the variable to  access in the action functions and read-write copy of the variable to access in the state functions. I do not know (because I just started learning to go) another way to provide read-only references. Maybe you can find a better solution.
Perhaps you need to put the state in the `current`

**Comment by breml** [comment]
@RedSkotina I will come back to you. I am currently on hollydays.

**Comment by breml** [comment]
@RedSkotina Back again.

My main points are the following:
1. The package global variable `state` has to go away, otherwise this PR will not get merged. The reason for this is, that the package global variable is shared between multiple instances of the parser (within the same program), which prevents them from being executed concurrently. This is a no-go.
2. I do not see the need of having `state` as a read-only variable. The user provided code, which uses the parser / is generated with the parser lives by design in the same package and therefore has full access to the internal state. So I see no need for the extra complexity as well as for the copying the state multiple times.
3. The new code block (`#{ }`) is not really needed (it is more like syntactic sugar). I would like to start with an implementation of the global state feature without this code block, where the state is updated via the and-/not-predicate code blocks.

Is it possible for you to change the PR in this direction (and rebase it onto current master)?

**Comment by RedSkotina** [comment]
@breml ok. i will begin work when i finish current project .(1-2 week)

**Comment by RedSkotina** [comment]
@breml I change PR as you ask and rebase it to master. It remains open to question some of the details. Memoization is one from them.

P.S. I has problem with windows ending of `static_code.go`, `static_code_range_table.go` and will create issue for it.

File: examples/indentation/indentation_test.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+			t.Errorf("%s: pigeon.ParseFile: %v", file, err)
+			continue
+		}
+        got,err := pgot.(ProgramNode).exec()
+        if err != nil {
+			t.Errorf("%s: ProgramNode.exec: %v", file, err)
+			continue
+		}
+        
+        exp := 42
```
Comment (breml) [inline examples/indentation/indentation_test.go]: On one hand, the test code is prepared to process several files from the `testdata` directory, on the other hand, the expected result is always 42. Maybe this could be optimized, e.g. by adding a `map[filename]result`.

File: builder/static_code.go [hunk 2]
```diff
@@ -136,10 +136,14 @@ type current struct {
 	pos  position // start position of the match
 	text []byte   // raw text of the match
 
+	// the state allows the parser to store arbitrary values and rollback them if needed
+	state statedict
 	// the globalStore allows the parser to store arbitrary values
 	globalStore map[string]interface{}
 }
 
+type statedict map[string]interface{}
```
[the remark this answers is not in the document]
Comment (RedSkotina) [reply builder/static_code.go]: no  advantage. just for clear code.

File: builder/static_code.go [hunk 3]
```diff
@@ -483,6 +488,33 @@ func (p *parser) restore(pt savepoint) {
...
+}
+
+// restore parser current state to the state statedict.
+func (p *parser) restoreState(state statedict) {
+	// ==template== {{ if not .Optimize }}
+	if p.debug {
+		defer p.out(p.in("restoreState"))
+	}
+	// {{ end }} ==template==
+	p.cur.state = make(statedict)
```
[the remark this answers is not in the document]
Comment (RedSkotina) [reply builder/static_code.go]: If you ask about restore state in restore() function than my answer: restore() is used in many other places. i will not restore state every time when backtrack token which cant change state (literals and etc)
If you ask about line #513 then you are right. copy-paste technology :)

File: doc.go [hunk 4]
```diff
@@ -311,6 +311,14 @@ runes from the start of the line, and offset is a 0-based byte offset.
 The "text" field is the slice of bytes of the current match. It is empty
 in a predicate code block.
 
+The "store" field is a global store with backtrack support of type "map[string]interface{}",
+which allows to store arbitrary values, which are available in action and
+predicate code blocks for read as well as write access.
+It is important to notice, that the global store is dependent from
+the backtrack mechanism of PEG and is can be set back to its old state
```
[the remark this answers is not in the document]
Comment (RedSkotina) [reply doc.go]: ok.

File: doc.go [hunk 5]
```diff
@@ -311,6 +311,14 @@ runes from the start of the line, and offset is a 0-based byte offset.
...
+during backtrack.
+Changes of "store" field in predicate expressions is not keeped.
```
[the remark this answers is not in the document]
Comment (RedSkotina) [reply doc.go]: I wanted to write that the state can not be changed inside and-not expressions ( only in and-not expresssion codebocks). E.g.
a -> point1 &Expr point2
I keep state in point 2 same like in point 1. Reason: &Expr just lookup symbols but not consume it. If we change state in Expr than we cant backtrack it is easy. Fix me if i wrong.

File: test/state/state.peg [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+}
+
+start = &{c.state["countCs"] = 0; return true,nil } ((x/y) ws*)* { return c.state["countCs"],nil }
+
+x = "ab" c "d"
+y = "a" bc "e"
+
+c = "c" &{ c.state["countCs"] = c.state["countCs"].(int)+3; return true,nil }
+bc = "bc" &{ c.state["countCs"] = c.state["countCs"].(int)+1; return true,nil}
```
[the remark this answers is not in the document]
Comment (RedSkotina) [reply test/state/state.peg]: ok

File: examples/indentation/indentation.peg [hunk 7]
```diff
@@ -0,0 +1,274 @@
+{
```
[the remark this answers is not in the document]
Comment (RedSkotina) [reply examples/indentation/indentation.peg]: ok

File: examples/indentation/indentation_test.go (hunk 1 above)
Comment (RedSkotina) [reply examples/indentation/indentation_test.go]: ok.

File: Makefile [hunk 8]
```diff
@@ -106,6 +109,10 @@ $(TEST_DIR)/global_store/global_store.go: $(TEST_DIR)/global_store/global_store.
 $(TEST_DIR)/goto/goto.go: $(TEST_DIR)/goto/goto.peg $(BINDIR)/pigeon
 	$(BINDIR)/pigeon $< > $@
 
+$(TEST_DIR)/state/state.go: $(TEST_DIR)/state/state.peg $(BINDIR)/pigeon
+	$(BINDIR)/pigeon $< | goimports > $@
+
+
```
[the remark this answers is not in the document]
Comment (RedSkotina) [reply Makefile]: ok. what about goimports?

**Comment by RedSkotina** [comment]
@breml. i use windows OS. so i have '\r\n' ending in files. it can be reason what we take different results.

**Comment by RedSkotina** [comment]
I remove all copyState, restoreState pairs and add only in `parseChoiceExpr`.
```
--- FAIL: TestState (0.00s)
        state_test.go:21: "abcf": want 5, got 6
```

File: Makefile (hunk 8 above)
[the remark this answers is not in the document]
Comment (breml) [reply Makefile]: goimports is no longer needed, as this functionality is now built into pigeon directly.

File: doc.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (breml) [reply doc.go]: Ok, now I understand, what you mean and I think this is an interesting question. I have to do some more thinking about the correct behavior. I will come back to you.

**Comment by breml** [comment]
@RedSkotina Thanks for the changes, they look good. I have to think again about  predicate expressions.

Regarding the windows style `\r\n` I prefere to keep the whole git repository in the same format so I ask you to also use unix style line endings (`\n`). Could you please convert your files (change the setting in your editor or configure git to use unix style line endings). Thanks.

**Comment by RedSkotina** [comment]
@breml  i found reason why tests failed when i test clone-restore pair in parseChoice only.
problem in new restoreState where we just assign p.cur.state = state
https://play.golang.org/p/Q_DWcUTLQe
When we enter to second choice `y` than we modidy our local copy `state` too, because `p.cur.state` share same dict with `state`.
so i will revert changes in restoreState

**Comment by RedSkotina** [comment]
@breml i added .gitattribute with forcing unix line endings.
can you recheck PR ?

**Comment by RedSkotina** [comment]
can you recheck PR? 
but i cant force my editor using lf only for one project. i need crlf for other my projects. 
seems better use some commit hook for checking line endings.(but i not so familiar with hooks)

File: .gitattributes [hunk 9]
```diff
@@ -0,0 +1,2 @@
+* text eol=lf
+* text eol=lf
```
Comment (breml) [inline .gitattributes]: Why do you add this line two times? isn't this redundant?

File: builder/static_code.go [hunk 10]
```diff
@@ -846,13 +878,15 @@ func (p *parser) parseChoiceExpr(ch *choiceExpr) (interface{}, bool) {
 	}
 
 	// {{ end }} ==template==
+	state := p.cloneState()
```
Comment (breml) [inline builder/static_code.go]: I agree, if you have `state := p.cloneState()` outside the for-loop, then you need the copying in `restoreState()`. My solution was to place `cloneState` inside the following for-loop.

File: examples/indentation/indentation_ast.go [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+	ret, err := pn.(ProgramNode).exec()
+	if err != nil {
+		log.Fatal(err)
+	}
+	fmt.Println(ret)
+}
+
+var lvalues = make(map[string]int)
+
+// element of statements
```
Comment (breml) [inline examples/indentation/indentation_ast.go]: I suggest to use unexported types here, as this is only example code. Otherwise `make lint` complains about the exported types not being commented properly:

```
examples/indentation/indentation_ast.go:34:1: comment on exported type Executor should be of the form "Executor ..." (with optional leading article)
```

This is true for all the exported types in this example.

File: .gitattributes (hunk 9 above)
Comment (RedSkotina) [reply .gitattributes]: seems wrong information from SO (first line for checkout, second for checkin). I am read now manual page and you right. Nothing about double lines.

File: builder/static_code.go (hunk 10 above)
Comment (RedSkotina) [reply builder/static_code.go]: ok. i revert it to your solution. And i add comment with restricts. Maybe we should use run-time checks for restrictions? Or this is useless increase of complexity?

File: Makefile (hunk 8 above)
[the remark this answers is not in the document]
Comment (RedSkotina) [reply Makefile]: removed.

File: examples/indentation/indentation_ast.go (hunk 11 above)
Comment (RedSkotina) [reply examples/indentation/indentation_ast.go]: I just expand all comments

**Comment by RedSkotina** [comment]
@breml what about states in And/Not Expr ?

File: .gitattributes [hunk 12]
```diff
@@ -0,0 +1 @@
+*.go text eol=lf
```
Comment (breml) [inline .gitattributes]: Why `lf` only for `*.go` files? I would like to have `lf` for all text files (especially but not limited to: `*.peg`, `*.md`).

**Comment by breml** [comment]
@RedSkotina I am aware, that the question regarding And/Not predicates is still open. I am still evaluating, the best way to go.

**Comment by breml** [comment]
@RedSkotina Do you still have the code with the `#{}` language element in your local repository? would you mind to push this code again as a separate PR?

**Comment by RedSkotina** [comment]
@breml yes. i have it pre rebase and after rebase. i need only merge it with current changes.

File: .gitattributes (hunk 12 above)
Comment (RedSkotina) [reply .gitattributes]: yes. i miss `*.peg`
because git sometime can change endline in binaries files  when used `* text eol=lf` (it is specified in manual). autodetect can work incorrectly.
lf endlines important only in `static_code.go` , `static_code_range_table.go` , because  it is directly read by bootstrap parser, and in `*.peg` files, because parser save position for them in grammar.
Other text files autoconverted from crlf to lf  after commit with windows git  standart settings .

**Comment by breml** [comment]
@RedSkotina Thanks for the updates. Everything is fine now, except for the open question, how `And`, `Not`, `AndCode` and `NotCode` expressions should behave in the case of backtracking.
Seriously I was thinking about this question yesterday as well as today and I also exchanged with a colleague about this question to find the best solution, but I am not yet done with it.

In fact, I am no longer sure, that it was a good idea (I know, it was my suggestion), to go with the `&{}` to change the state and I am also thinking again about the option to introduce a new language element (like the `#{}` you proposed in the first PR).

This being said, I also consider the impact of this decision on the following to features:

* global store, already present in pigeon, where I currently suggest to use the `&{}`/AndCode expression
* labeled failures (https://github.com/breml/pigeon/blob/labeled-failures/test/labeled_failures/labeled_failures.peg#L30-L35), a feature, which I am working on, and where currently also the `&{}`/AndCode expressions are used to record failures during parsing, without backtracking.

At the moment I am considering several options, which takes some time. Sorry for the delay.

**Comment by breml** [comment]
@RedSkotina I have not forgotten you. I hope I can provide my proposal by the end of this week.

**Comment by breml** [comment]
@RedSkotina After quite some thinking, I ended up with the proposal in #40, which describes, how I think, the **global state** and **global store** should be implemented. It is the result of our intense discussion in your [PR](https://github.com/mna/pigeon/pull/16), especially the question about how the `AndExpr` and `NotExpr` should behave (see [comment](https://github.com/mna/pigeon/pull/16#issuecomment-325504426)), as well as some discussion I had with colleagues about these features.

Currently there are still some questions open, especially in regards to the errors in case of backtracking.

Of course, this proposal is open for discussion and I am happy to answer your question about it.

**Comment by breml** [comment]
@RedSkotina What do you think about the current situation? Do you feel like extending this PR on the basis of https://github.com/mna/pigeon/issues/40?

**Comment by RedSkotina** [comment]
@breml  i will answer after one week. sorry,  not have time right now.

**Comment by breml** [comment]
@RedSkotina I made an initial implementation of #40 based on this PR, have a look at #43.

**Comment by breml** [comment]
@RedSkotina Your comments have been integrated into #43, which has been merged to master. Thanks for your helpful work to make this feature happen.
